From 87fdcc5f0e53b6dcdcfc8dc0b10a3ebe662c9c35 Mon Sep 17 00:00:00 2001
From: Alan Pevec <apevec@redhat.com>
Date: Mon, 9 Sep 2013 18:58:47 +0200
Subject: [PATCH] Use updated parallel install versions of epel package

Use SQLAlchemy >= 0.7.8 WebOb >= 1.2 Routes >= 1.12.3 PasteDeploy >= 1.5.0
and depend on the parallel installable
versions of these packages to satisfy those requirements.
---
 bin/keystone-all     |  4 ++--
 keystone/__init__.py | 29 +++++++++++++++++++++++++++++
 2 files changed, 31 insertions(+), 2 deletions(-)

diff --git a/bin/keystone-all b/bin/keystone-all
index 02294d2..8f1fe89 100755
--- a/bin/keystone-all
+++ b/bin/keystone-all
@@ -31,10 +31,10 @@ if os.path.exists(os.path.join(possible_topdir,
     sys.path.insert(0, possible_topdir)
 
 
-from paste import deploy
-
 from keystone.openstack.common import gettextutils
 
+from paste import deploy
+
 # NOTE(blk-u):
 # gettextutils.install() must run to set _ before importing any modules that
 # contain static translated strings.
diff --git a/keystone/__init__.py b/keystone/__init__.py
index e69de29..6af9f66 100644
--- a/keystone/__init__.py
+++ b/keystone/__init__.py
@@ -0,0 +1,29 @@
+import sys
+import pkg_resources
+
+# If there is a conflicting non egg module,
+# i.e. an older standard system module installed,
+# then replace it with this requirement
+def replace_dist(requirement):
+    try:
+        return pkg_resources.require(requirement)
+    except pkg_resources.VersionConflict:
+        e = sys.exc_info()[1]
+        dist=e.args[0]
+        req=e.args[1]
+        if dist.key == req.key and not dist.location.endswith('.egg'):
+            del pkg_resources.working_set.by_key[dist.key]
+            # We assume there is no need to adjust sys.path
+            # and the associated pkg_resources.working_set.entries
+            return pkg_resources.require(requirement)
+
+replace_dist("WebOb >= 1.2")
+replace_dist("SQLAlchemy >= 0.7.8")
+replace_dist("Routes >= 1.12.3")
+
+replace_dist("PasteDeploy >= 1.5.0")
+# This hack is needed because replace_dist() results in
+# the standard paste module path being at the start of __path__.
+# TODO: See can we get pkg_resources to do the right thing directly
+import paste
+paste.__path__.insert(0, paste.__path__.pop(-1))
